source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -0,0 +1,305 @@
+#include "extensions/filters/udp/dns_filter/dns_parser.h"
+
+#include <iomanip>
+#include <sstream>
+
+#include "envoy/network/address.h"
+
+#include "common/common/empty_string.h"
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace DnsFilter {
+
+inline void BaseDnsRecord::serializeName() {
+
+  // Iterate over a name e.g. "www.domain.com" once and produce a buffer containing each name
+  // segment prefixed by its length
+
+  static constexpr char SEPARATOR('.');
+
+  size_t last = 0;
+  size_t count = name_.find_first_of(SEPARATOR);
+  auto iter = name_.begin();
+
+  while (count != std::string::npos) {
+
+    count -= last;
+    buffer_.writeBEInt<uint8_t>(count);
+    for (size_t i = 0; i < count; i++) {
+      buffer_.writeByte(*iter);
+      ++iter;
+    }
+
+    // periods are not serialized. Skip to the next character
+    if (*iter == SEPARATOR) {
+      ++iter;
+    }
+
+    // Move our last marker to the first position after where we stopped. Search for the next name
+    // separator
+    last += count;
+    ++last;
+    count = name_.find_first_of(SEPARATOR, last);
+  }
+
+  // Write the remaining segment prepended by its length
+  count = name_.size() - last;
+  buffer_.writeBEInt<uint8_t>(count);
+  for (size_t i = 0; i < count; i++) {
+    buffer_.writeByte(*iter);
+    ++iter;
+  }
+
+  // Terminate the name record with a null byte
+  buffer_.writeByte(0x00);
+}
+
+// Serialize a DNS Query Record
+void DnsQueryRecord::serialize(Buffer::OwnedImpl& output) {
+  buffer_.drain(buffer_.length());
+
+  serializeName();
+  buffer_.writeBEInt<uint16_t>(type_);
+  buffer_.writeBEInt<uint16_t>(class_);
+
+  output.add(buffer_);
+}
+
+bool DnsMessageParser::parseDnsObject(const Buffer::InstancePtr& buffer) {
+
+  auto available_bytes = buffer->length();
+
+  memset(&incoming_, 0x00, sizeof(incoming_));
+
+  static constexpr uint64_t field_size = sizeof(uint16_t);
+  uint64_t offset = 0;
+  uint16_t data;
+
+  DnsQueryParseState state_{DnsQueryParseState::Init};
```

Comment:
the name is just "state" as its local var?

Reply:
ping on this comment and below comments. Please resolve all comments that are actioned on.

Reply:
will do